frontend-project/src/pages/events/EventsDetailView.tsx

```diff
@@ -0,0 +1,235 @@
+import { PageHeaderWrapper } from '@ant-design/pro-layout';
+import { Button, Col, Card, Form, Input, Row, Select, Space, Spin } from 'antd';
+import { connect, useDispatch } from 'dva';
+import React, { useEffect, useState } from 'react';
+import { formatMessage, FormattedMessage } from 'umi-plugin-react/locale';
+import { Event, Case } from '@/services/definitions';
+import { ReduxResourceState } from '@/utils/reduxModel';
+import router from 'umi/router';
+import { localeKeys } from '../../locales/pl-PL';
+import { RouterTypes } from 'umi';
+import { ServiceResponse } from '@/services/service';
+import { openNotificationWithIcon } from '@/models/global';
+
+interface EventsDetailViewProps {
+  events: ReduxResourceState<Event>;
+  cases: ReduxResourceState<Case>;
+  match: RouterTypes['match'] & { params: { id: string | undefined } };
+}
+
+const { TextArea } = Input;
+const { Option } = Select;
+
+const layout = {
+  labelCol: { span: 8 },
+  wrapperCol: { span: 16 },
+};
+
+const tailLayout = {
+  wrapperCol: { offset: 8, span: 16 },
+};
+
+function EventsDetailView({ events, cases, match }: EventsDetailViewProps) {
+  const dispatch = useDispatch();
+  const isEdit = Boolean(match.params.id);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const editedEvent = events.data.find(value => value.id === Number(match.params.id));
+  const [form] = Form.useForm();
+  function onRequestDone(response: ServiceResponse<Event>) {
```

Comment:
Robi się trochę WET: funkcje `onRequestDone` i `onFinish` są kopiowane-wklejane z komponentu na komponent z paroma tylko modyfikacjami (`router.push({ where })`, `localeKeys.cases.{ whichView }`).

Warto zrobić w tym momencie refactor, jakoś sparametryzować, wyciągnąć do osobnego pliku, tak, żeby widok mógł ich używać jakoś tak:
```
const onRequestDone = onRequestDoneFactory<Event>(form, '/cases', 'detailView');

const onFinish = onFinishFactory('cases', onRequestDone);
```
Czy coś w tym guście.

```

Reply:
Zarówno onRequestDone jak i onFinish zależą w dużej mierze od kontekstu (onFInish: nazwa resourca, stan isSubitting, dispatch, form i opcjonalny resoureId, onRequestDone: nazwa resource, form, router, locale, stan isSubmitting). Wyciąganie tego do generyków spowoduje że będziemy mieli 2x funkcje która przyjmuje 5 albo jedną funkcję która przyjmuje 7 parametrów. Troszeczkę popracowałem z localami żeby były bardziej zwięzłe, ale wydaje mi się że to są zbyt zależne od kontekstu funkcje żeby je przenosić do generyków.

Reply:
To nie wymaga aż tak znowu wielu parametrów: np. `router` to zwykły import. Dla `onRequestDone` będzie tego cztery parametry (setIsSubmitting, form, 'localeKeys.users', isEdit), z `onFinish` jest istotnie gorzej. Można zacząć od `onRequestDone`, `onFinish` odkładając na następny raz.

Natomiast takie kopiuj-wklej skutkuje kłopotami w przyszłości (gdy np. będzie potrzeba zmiany zachowania aplikacji - trzeba będzie tę samą zmianę pieczołowicie wydłubać w każdym komponencie), i prowokuje bugi (np. teraz komunikat błędu w `UsersListView` to "Nie udało się usunąć sprawy" - `formatMessage({ id: localeKeys.cases.list.failedRemove })` - a nie "...usunąć użytkownika"). Im większy kawał kodu kopiujesz-wklejasz-zmieniasz tylko istotne kawałki, tym łatwiej takie błędy powstają.

Reply:
@fervero Zmiany do wszystkich widoków wrzucę w osobnym PR, bo obecnie znacząco rozrósł by się.